lib/class-wp-rest-nav-menu-locations-controller.php

```diff
@@ -0,0 +1,369 @@
+<?php
+/**
+ * REST API: WP_REST_Nav_Menu_Locations_Controller class
+ *
+ * @package WordPress
+ * @subpackage REST_API
+ * @since 5.x
+ */
+
+/**
+ * Core controller used to access menu locations via the REST API.
+ *
+ * @since 5.x
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_Nav_Menu_Locations_Controller extends WP_REST_Controller {
+
+	/**
+	 * The namespace of this controller's route.
+	 *
+	 * @since 5.x
+	 *
+	 * @var string
+	 */
+	protected $namespace = 'wp/v2';
+
+	/**
+	 * The base of this controller's route.
+	 *
+	 * @since 5.x
+	 *
+	 * @var string
+	 */
+	protected $rest_base = 'menu-locations';
+
+	/**
+	 * Registers the routes for the objects of the controller.
+	 *
+	 * @since 5.x
+	 *
+	 * @see register_rest_route()
+	 */
+	public function register_routes() {
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+					'args'                => $this->get_collection_params(),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<location>[\w-]+)',
+			array(
+				'args'   => array(
+					'location' => array(
+						'description' => __( 'An alphanumeric identifier for the menu location.' ),
+						'type'        => 'string',
+					),
+				),
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_item' ),
+					'permission_callback' => array( $this, 'get_item_permissions_check' ),
+					'args'                => array(
+						'context' => $this->get_context_param( array( 'default' => 'view' ) ),
+					),
+				),
+				array(
+					'methods'             => WP_REST_Server::EDITABLE,
+					'callback'            => array( $this, 'update_item' ),
+					'permission_callback' => array( $this, 'update_item_permissions_check' ),
+					'args'                => array(
+						'menu' => array(
+							'validate_callback' => function( $id ) {
+								return 0 === $id || false !== wp_get_nav_menu_object( $id );
+							},
+							'required'          => true,
+						),
+					),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Checks whether a given request has permission to read menu locations.
+	 *
+	 * @since 5.x
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|bool True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) {
+		if ( ! current_user_can( 'edit_theme_options' ) ) {
```

Comment:
I think that permission checks should allow the post type's permissions.

Reply:
This controller doesn't handle a post type. It is the locations controller and allows you to reassign a nav menu to a registered menu location.